### ISSUE TITLE
Suggestion: Emit warning when fitsio.write doesn't overwrite a file since clobber=False.

### ISSUE THREAD

**Issue by rmjarvis** [esheldon/fitsio#336]
I've been bitten by this fitsio feature multiple times: When calling `fitsio.write(file_name, data)`, if the file exists already, fitsio will append a new extension rather than overwrite the file.

I get that this is intended behavior, and is correctly documented.  There is also a simple parameter to change it `clobber=True`.

However, I think the UI could be improved by adding a warning when this occurs, since if you're expecting the file to be overwritten, it seems like it did nothing, since the first, default extension is unchanged.  So reading in that file somewhere else ends up with the same, old information, not the new data you meant to write to the file.

Therefore, I'm suggesting that when clobber=False and the file exists, fitsio emit a warning letting the user know fitsio is doing, since I'm guessing I'm not the only user who finds this a bit counter-intuitive.

**Comment by mickolaua** [comment]
> I've been bitten by this fitsio feature multiple times: When calling fitsio.write(file_name, data), if the file exists already, fitsio will append a new extension rather than overwrite the file.

Encountered the same issue, noticed it after couple of runs when the file size increased.

**Comment by esheldon** [comment]
This is default cfitsio behavior, and is also the behavior of mwrfits from IDL.  I think it is terrible but I was trying to stick with what people would expect.

Now that the user base is mostly people who never used either cfitsio or mwrfits, it is surprising!  There is no win in these situations.

The problem with the warning is when someone does want to add a new extension (not uncommon!) they will get crap in their logs.

**Comment by rmjarvis** [comment]
I think usually when you would be intentionally adding an extension, you'd probably name it.  (Not required of course, but it is good practice.)  So maybe only emit a warning if clobber=False AND extname is None?

Or possible even better, have the default clobber be None, rather than False, and then only emit a warning if clobber is None and extname is None with part of the message saying that explicitly using clobber=False will silence the warning.

**Comment by rmjarvis** [comment]
(Good API design is hard...)

**Comment by ntessore** [comment]
I just had a big argument with this one. I would like to propose raising an exception by default if a file exists for writing. This would probably be the path of least unexpected behaviour for most people.

In combination with that, the default could be `clobber=None` which raises the exception. The user can then specifically set `clobber=False` to force appending, or `clobber=True` to force clobbering.

Even better would be to introduce new explicit file modes `r`, `w`, `a` in both the FITS opener and the convenience routines (and keep `rw` for the current behaviour). Then we could say very specifically what we want the code to do (explicit is better than implicit), e.g. `fitsio.write(file, data, mode='a')`.

**Comment by esheldon** [comment]
How about we add a new open mode for fits files.  opening with `"w"` would always clobber.

```python
# opening with "w"
with fitsio.FITS(filename, 'w') as fits:
    fits.write(data)

# would be equivalent of
with fitsio.FITS(filename, 'rw', clobber=True) as fits:
    fits.write(data)
```

Or we could entertain other strings besides `"w"`

This doesn't fix the convenience function `fitsio.write(filename, data)` but would give a more standard "clobber" interface for the context

**Comment by esheldon** [comment]
Maybe `"w+"`

**Comment by esheldon** [comment]
Sorry @ntessore I see you also suggested some such things.  I'd like to reopen the discussion

**Comment by ntessore** [comment]
I still think this would be a very welcome feature. But perhaps now I would suggest deprecating/removing the `clobber=` argument entirely, and using a mode string that mirrors the behaviour of Python's [`open()`](https://docs.python.org/3/library/functions.html#open):

* `"r"` open for reading (default), failing if file does not exist
* `"w"` open for writing, truncating the file first
* `"x"` open for exclusive creation, failing if the file already exists 
* `"a"` open for writing, appending to the file if it exists

And since fitsio does (AFAIK) not do write-only opening, anyway, the writing modes would naturally support reading as well.

**Comment by beckermr** [comment]
So the `clobber` stuff is part of cfitsio itself, not something invented by `fitsio`. We thus should not remove it. Supply a more standard set of IO flags when dealing directly with file pointers is a good compromise.

**Comment by ntessore** [comment]
Then why not a new helper function `fitsio.open()`, say, that first does the potential existence checks and sets `clobber=` on the FITS instance as required?

```py
with fitsio.open("myfile.fits", "x"):
    ...
```

I think this could be a pretty comfortable pythonic interface.

**Comment by beckermr** [comment]
Sure or we can embed this behavior directly in FITS itself. My only point was that we should not remove the `clobber` keyword argument.

**Comment by ntessore** [comment]
Makes sense! I was thinking more about not having to decide how to deal with mixed-up inputs like `FITS(..., "a", clobber=True)`.

**Comment by esheldon** [comment]
clobber is everywhere in the wild, I can't even speculate how many  lines of code must have that.

I don't want to break any existing code, or force people to change through deprecation ("we do not break userspace" - Linus)

For this reason I want to consider new APIs.

I'll expand on @ntessore list here:

* `r` would continue as it works now. `FITS(fname)`

* `w` For consistency with python, we might want to allow both `w+` and `w` although the behavior would be the same.   These would be equivalent to our current `FITS(fname, 'rw', clobber=True)`.

* `a` (and maybe also `a+`) would be equivalent to our current `FITS(fname, 'rw')` 

* `x` is a new thing, and would behave as @ntessore says

We could disallow  use of `clobber` with a mode that doesn't make sense.

The convenience function `write` is not going to be easy to fix, but we could add new ones.  Maybe

```
fitsio.create(fname, data)  # equivalent of fitsio.write(fname, data, clobber=True)
fitsio.append(fname, data)  # equivalent of fitsio.write(fname, data) but demands file exists
```

**Comment by beckermr** [comment]
That all looks fine. I'd personally not create new convenience functions myself.
